File: cc/016.cc
#include <iostream>
using namespace std;

class ListNode {
public:
    int val;
    ListNode * next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* Merge(ListNode* pHead1, ListNode* pHead2)
    {
        if (!pHead1) return pHead2;
        if (!pHead2) return pHead1;
        ListNode * pHead3 = pHead1->val < pHead2->val ? pHead1 : pHead2;
        pHead3->next = pHead1->val < pHead2->val ? this->Merge(pHead1->next, pHead2) : this->Merge(pHead1, pHead2->next);
        return pHead3;
    }
};

int main(int argc, char const *argv[])
{
    int vals1[] = {1, 2, 3, 5, 7, 9};
    ListNode * l1 = new ListNode(vals1[0]);
    ListNode * p = l1;
    for (int i = 1; i < 6; i++)
    {
        p->next = new ListNode(vals1[i]);
        p = p->next;
    }
    int vals2[] = {2, 4, 5, 6, 8, 10};
    ListNode * l2 = new ListNode(vals2[0]);
    p = l2;
    for (int i = 1; i < 6; i++)
    {
        p->next = new ListNode(vals2[i]);
        p = p->next;
    }
    Solution * s = new Solution();
    p = s->Merge(l1, l2);
    while (p)
    {
        cout << p->val << endl;
        p = p->next;
    }
    return 0;
}

File: python/016.py
# -*- coding: utf-8 -*-
# /usr/bin/python3
# Creator: Hemajun
# Description: 输入两个单调递增的链表，输出两个链表合成后的链表，当然我们需要合成后的链表满足单调不减规则。


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # 返回合并后列表
    def Merge(self, pHead1, pHead2):
        head = ListNode(0)
        p = head
        while pHead1 is not None and pHead2 is not None:
            if pHead1.val <= pHead2.val:
                p.next = pHead1
                pHead1 = pHead1.next
            else:
                p.next = pHead2
                pHead2 = pHead2.next
            p = p.next
        if pHead1 is not None:
            p.next = pHead1
        if pHead2 is not None:
            p.next = pHead2
        return head.next
